File: packages/sdk/src/derivation.ts
import { JUP_STABLE_PROGRAM_ADDRESS } from "./generated";
import {
  Address,
  getAddressEncoder,
  getBytesEncoder,
  getProgramDerivedAddress,
} from "@solana/kit";
import {
  findAssociatedTokenPda,
  TOKEN_PROGRAM_ADDRESS,
} from "@solana-program/token";

export async function findConfig(): Promise<Address> {
  const [configAccount] = await getProgramDerivedAddress({
    programAddress: JUP_STABLE_PROGRAM_ADDRESS,
    seeds: [getBytesEncoder().encode(Buffer.from("config"))],
  });
  return configAccount;
}

export async function findAuthority(): Promise<Address> {
  const [authorityAccount] = await getProgramDerivedAddress({
    programAddress: JUP_STABLE_PROGRAM_ADDRESS,
    seeds: [getBytesEncoder().encode(Buffer.from("authority"))],
  });
  return authorityAccount;
}

export async function findOperator(authority: Address): Promise<Address> {
  const [operatorAccount] = await getProgramDerivedAddress({
    programAddress: JUP_STABLE_PROGRAM_ADDRESS,
    seeds: [
      getBytesEncoder().encode(Buffer.from("operator")),
      getAddressEncoder().encode(authority),
    ],
  });
  return operatorAccount;
}

export async function findBenefactor(authority: Address): Promise<Address> {
  const [benefactorAccount] = await getProgramDerivedAddress({
    programAddress: JUP_STABLE_PROGRAM_ADDRESS,
    seeds: [
      getBytesEncoder().encode(Buffer.from("benefactor")),
      getAddressEncoder().encode(authority),
    ],
  });
  return benefactorAccount;
}

export async function findVault(mint: Address): Promise<Address> {
  const [vaultAccount] = await getProgramDerivedAddress({
    programAddress: JUP_STABLE_PROGRAM_ADDRESS,
    seeds: [
      getBytesEncoder().encode(Buffer.from("vault")),
      getAddressEncoder().encode(mint),
    ],
  });
  return vaultAccount;
}

export async function findVaultTokenAccount(mint: Address): Promise<Address> {
  const [vaultTokenAccount] = await findAssociatedTokenPda({
    mint: mint,
    owner: await findAuthority(),
    tokenProgram: TOKEN_PROGRAM_ADDRESS,
  });
  return vaultTokenAccount;
}

File: programs/jup-stable/tests/common/derivation.rs
use anchor_spl::metadata;
use solana_sdk::pubkey::Pubkey;
use spl_associated_token_account::get_associated_token_address_with_program_id;

pub fn find_config() -> Pubkey {
    let (pubkey, _bump) = Pubkey::find_program_address(&[b"config"], &jup_stable::id());
    pubkey
}

pub fn find_authority() -> Pubkey {
    let (pubkey, _bump) = Pubkey::find_program_address(&[b"authority"], &jup_stable::id());
    pubkey
}

pub fn find_operator(authority: &Pubkey) -> Pubkey {
    let (pubkey, _bump) =
        Pubkey::find_program_address(&[b"operator", authority.as_ref()], &jup_stable::id());
    pubkey
}

pub fn find_vault(stablecoin_mint: &Pubkey) -> Pubkey {
    let (pubkey, _bump) =
        Pubkey::find_program_address(&[b"vault", stablecoin_mint.as_ref()], &jup_stable::id());
    pubkey
}

pub fn find_vault_token_account(stablecoin_mint: &Pubkey) -> Pubkey {
    get_associated_token_address_with_program_id(&find_authority(), stablecoin_mint, &spl_token::ID)
}

pub fn find_benefactor(authority: &Pubkey) -> Pubkey {
    let (pubkey, _bump) =
        Pubkey::find_program_address(&[b"benefactor", authority.as_ref()], &jup_stable::id());
    pubkey
}

pub fn find_metadata(mint: &Pubkey) -> Pubkey {
    let (pubkey, _bump) = Pubkey::find_program_address(
        &[b"metadata", &metadata::ID.to_bytes(), &mint.to_bytes()],
        &metadata::ID,
    );
    pubkey
}

pub fn find_event_authority() -> Pubkey {
    let (pubkey, _bump) = Pubkey::find_program_address(&[b"__event_authority"], &jup_stable::id());
    pubkey
}
